File: Controls/ResizableCanvas.vb
Imports System.Windows.Media
Imports Ks.Common

Namespace Controls

    Public Class ResizableCanvas
        Inherits Panel

        Private Function GetViewingArea(ByVal Size As Size) As Rect?
            Dim ViewingRectangle = Me.ViewingRectangle
            'ViewingRectangle.Intersect(ShapeRectangle)

            If Me.KeepAspectRatio Then
                Dim RB = ViewingRectangle.GetSmallestBoundOf(Me.GetClientArea(Size).Size)
                If Not RB.Item2.HasValue Then
                    Console.WriteLine("Error. " & Utilities.Debug.CompactStackTrace(5))
                    Return Nothing
                End If
                ViewingRectangle = RB.Item1
                'If Not ShapeRectangle.Contains(ViewingRectangle) Then
                '    If RB.Item2.Value Then
                '        Dim B1 = ViewingRectangle.Top < ShapeRectangle.Top
                '        Dim B2 = ViewingRectangle.Bottom > ShapeRectangle.Bottom
                '        If B1 Xor B2 Then
                '            If B1 Then
                '                ViewingRectangle.Y = ShapeRectangle.Y
                '            Else
                '                ViewingRectangle.Y = ShapeRectangle.Y + ShapeRectangle.Height - ViewingRectangle.Height
                '            End If
                '        End If
                '    Else
                '        Dim B1 = ViewingRectangle.Left < ShapeRectangle.Left
                '        Dim B2 = ViewingRectangle.Right > ShapeRectangle.Right
                '        If B1 Xor B2 Then
                '            If B1 Then
                '                ViewingRectangle.X = ShapeRectangle.X
                '            Else
                '                ViewingRectangle.X = ShapeRectangle.X + ShapeRectangle.Width - ViewingRectangle.Width
                '            End If
                '        End If
                '    End If
                'End If
                ''If Not ShapeRectangle.Contains(ViewingRectangle) Then
                ''    RB = ShapeRectangle.GetLargestFitOf(Client.Size)
                ''    ViewingRectangle = RB.Item1
                ''End If
            End If

            Return ViewingRectangle
        End Function

        Private Function GetClientArea(ByVal Size As Size) As Rect
            Dim Padding = Me.Padding
            Dim Padding1 = New Point(Padding.Left, Padding.Top)
            Dim Padding2 = New Point(Padding.Right, Padding.Bottom)

            Return New Rect(Padding1,
                            (-Padding2.ToVector() + Size.ToVector()).ToPoint())
        End Function

        Protected Overrides Function MeasureOverride(ByVal AvailableSize As Size) As Size
            Dim InfSize = New Size(Double.PositiveInfinity, Double.PositiveInfinity)

            Dim Client = Me.GetClientArea(AvailableSize)

            Dim ViewingRectangleQ = Me.GetViewingArea(AvailableSize)
            If Not ViewingRectangleQ.HasValue Then
                Return AvailableSize
            End If
            Dim ViewingRectangle = ViewingRectangleQ.Value

            Dim Sz0 = New Point()

            Sz0 = ViewingRectangle.ToLocal01(Sz0)
            Sz0 = Client.FromLocal01(Sz0)

            For Each C As UIElement In Me.Children
                Dim Sz = New Vector(GetW(C), GetH(C))

                If Sz = New Vector() Then
                    C.Measure(InfSize)
                Else
                    Dim Szz = Sz.ToPoint()

                    Szz = ViewingRectangle.ToLocal01(Szz)
                    Szz = Client.FromLocal01(Szz)

                    Dim Size = (Szz - Sz0).ToSize()
                    If Double.IsNaN(Size.Width) Or Double.IsNaN(Size.Height) Then
                        Size = New Size()
                    End If

                    C.Measure(Size)
                End If
            Next

            Return New Size()
        End Function

        Protected Overrides Function ArrangeOverride(FinalSize As Size) As Size
            Dim Client = Me.GetClientArea(FinalSize)

            Dim ViewingRectangleQ = Me.GetViewingArea(FinalSize)
            If Not ViewingRectangleQ.HasValue Then
                Return FinalSize
            End If
            Dim ViewingRectangle = ViewingRectangleQ.Value

            For Each C As UIElement In Me.Children
                Dim Pt = New Point(GetX(C), GetY(C))
                Dim Sz = New Vector(GetW(C), GetH(C))

                Dim Pt1 = Pt
                Pt1 = ViewingRectangle.ToLocal01(Pt1)
                Pt1 = Client.FromLocal01(Pt1)

                If Sz = New Vector() Then
                    C.Arrange(New Rect(Pt1, C.DesiredSize))
                Else
                    Dim Pt2 = Pt + Sz
                    Pt2 = ViewingRectangle.ToLocal01(Pt2)
                    Pt2 = Client.FromLocal01(Pt2)
                    C.Arrange(New Rect(Pt1, Pt2))
                End If
            Next

            Return FinalSize
        End Function

        ' ToDo Support Padding and MaxViewingRectangle.

#Region "ViewingRectangle Property"
        Public Shared ReadOnly ViewingRectangleProperty As DependencyProperty = DependencyProperty.Register("ViewingRectangle", GetType(Rect), GetType(ResizableCanvas), New PropertyMetadata(New Rect(0, 0, 1, 1), AddressOf ViewingRectangle_Changed))

        Private Shared Sub ViewingRectangle_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, ResizableCanvas)
            Self.InvalidateMeasure()
        End Sub

        Public Property ViewingRectangle As Rect
            Get
                Return DirectCast(Me.GetValue(ViewingRectangleProperty), Rect)
            End Get
            Set(value As Rect)
                Me.SetValue(ViewingRectangleProperty, value)
            End Set
        End Property
#End Region

        '#Region "MaxViewingRectangle Property"
        '        Public Shared ReadOnly MaxViewingRectangleProperty As DependencyProperty = DependencyProperty.Register("MaxViewingRectangle", GetType(Rect?), GetType(ResizableCanvas), New PropertyMetadata(New Rect?(), AddressOf MaxViewingRectangle_Changed))

        '        Private Shared Sub MaxViewingRectangle_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
        '            Dim Self = DirectCast(D, ResizableCanvas)

        '            Dim OldValue = DirectCast(E.OldValue, Rect?)
        '            Dim NewValue = DirectCast(E.NewValue, Rect?)
        '        End Sub

        '        Public Property MaxViewingRectangle As Rect?
        '            Get
        '                Return DirectCast(Me.GetValue(MaxViewingRectangleProperty), Rect?)
        '            End Get
        '            Set(ByVal value As Rect?)
        '                Me.SetValue(MaxViewingRectangleProperty, value)
        '            End Set
        '        End Property
        '#End Region

#Region "KeepAspectRatio Property"
        Public Shared ReadOnly KeepAspectRatioProperty As DependencyProperty = DependencyProperty.Register("KeepAspectRatio", GetType(Boolean), GetType(ResizableCanvas), New PropertyMetadata(True, AddressOf KeepAspectRatio_Changed))

        Private Shared Sub KeepAspectRatio_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, ResizableCanvas)
            Self.InvalidateMeasure()
        End Sub

        Public Property KeepAspectRatio As Boolean
            Get
                Return DirectCast(Me.GetValue(KeepAspectRatioProperty), Boolean)
            End Get
            Set(value As Boolean)
                Me.SetValue(KeepAspectRatioProperty, value)
            End Set
        End Property
#End Region

#Region "Padding Property"
        Public Shared ReadOnly PaddingProperty As DependencyProperty = DependencyProperty.Register("Padding", GetType(Thickness), GetType(ResizableCanvas), New PropertyMetadata(New Thickness(), AddressOf Padding_Changed))

        Private Shared Sub Padding_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, ResizableCanvas)
            Self.InvalidateMeasure()
        End Sub

        Public Property Padding As Thickness
            Get
                Return DirectCast(Me.GetValue(PaddingProperty), Thickness)
            End Get
            Set(value As Thickness)
                Me.SetValue(PaddingProperty, value)
            End Set
        End Property
#End Region

#Region "Y Property"
        Public Shared ReadOnly YProperty As DependencyProperty = DependencyProperty.RegisterAttached("Y", GetType(Double), GetType(ResizableCanvas), New PropertyMetadata(0.0, AddressOf Y_Changed))

        Private Shared Sub Y_Changed(ByVal D As System.Windows.DependencyObject, ByVal E As System.Windows.DependencyPropertyChangedEventArgs)
            Dim C = TryCast(VisualTreeHelper.GetParent(D), ResizableCanvas)
            If C IsNot Nothing Then
                C.InvalidateMeasure()
            End If
        End Sub

        Public Shared Function GetY(ByVal O As UIElement) As Double
            Return DirectCast(O.GetValue(YProperty), Double)
        End Function

        Public Shared Sub SetY(ByVal O As UIElement, ByVal Value As Double)
            O.SetValue(YProperty, Value)
        End Sub
#End Region

#Region "X Property"
        Public Shared ReadOnly XProperty As DependencyProperty = DependencyProperty.RegisterAttached("X", GetType(Double), GetType(ResizableCanvas), New PropertyMetadata(0.0, AddressOf X_Changed))

        Private Shared Sub X_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim C = TryCast(VisualTreeHelper.GetParent(D), ResizableCanvas)
            If C IsNot Nothing Then
                C.InvalidateMeasure()
            End If
        End Sub

        Public Shared Function GetX(ByVal O As UIElement) As Double
            Return DirectCast(O.GetValue(XProperty), Double)
        End Function

        Public Shared Sub SetX(ByVal O As UIElement, ByVal Value As Double)
            O.SetValue(XProperty, Value)
        End Sub
#End Region

#Region "H Property"
        Public Shared ReadOnly HProperty As DependencyProperty = DependencyProperty.RegisterAttached("H", GetType(Double), GetType(ResizableCanvas), New PropertyMetadata(0.0, AddressOf H_Changed))

        Private Shared Sub H_Changed(ByVal D As System.Windows.DependencyObject, ByVal E As System.Windows.DependencyPropertyChangedEventArgs)
            Dim C = TryCast(VisualTreeHelper.GetParent(D), ResizableCanvas)
            If C IsNot Nothing Then
                C.InvalidateMeasure()
            End If
        End Sub

        Public Shared Function GetH(ByVal O As UIElement) As Double
            Return DirectCast(O.GetValue(HProperty), Double)
        End Function

        Public Shared Sub SetH(ByVal O As UIElement, ByVal Value As Double)
            O.SetValue(HProperty, Value)
        End Sub
#End Region

#Region "W Property"
        Public Shared ReadOnly WProperty As DependencyProperty = DependencyProperty.RegisterAttached("W", GetType(Double), GetType(ResizableCanvas), New PropertyMetadata(0.0, AddressOf W_Changed))

        Private Shared Sub W_Changed(ByVal D As System.Windows.DependencyObject, ByVal E As System.Windows.DependencyPropertyChangedEventArgs)
            Dim C = TryCast(VisualTreeHelper.GetParent(D), ResizableCanvas)
            If C IsNot Nothing Then
                C.InvalidateMeasure()
            End If
        End Sub

        Public Shared Function GetW(ByVal O As UIElement) As Double
            Return DirectCast(O.GetValue(WProperty), Double)
        End Function

        Public Shared Sub SetW(ByVal O As UIElement, ByVal Value As Double)
            O.SetValue(WProperty, Value)
        End Sub
#End Region

    End Class

End Namespace

File: WPF/Controls/ResizableCanvas.cs
using System.Windows;
using System.Windows.Controls;
using System;
using System.Windows.Media;
using Ks.Common;

namespace Ks.Common.Controls
{
    public class ResizableCanvas : Panel
    {
        private Rect? GetViewingArea(Size Size)
        {
            var ViewingRectangle = this.ViewingRectangle;
            // ViewingRectangle.Intersect(ShapeRectangle)

            if (this.KeepAspectRatio)
            {
                var RB = ViewingRectangle.GetSmallestBoundOf(this.GetClientArea(Size).Size);
                if (!RB.Item2.HasValue)
                {
                    Console.WriteLine("Error. " + Utilities.Debug.CompactStackTrace(5));
                    return default;
                }
                ViewingRectangle = RB.Item1;
            }

            return ViewingRectangle;
        }

        private Rect GetClientArea(Size Size)
        {
            var Padding = this.Padding;
            var Padding1 = new Point(Padding.Left, Padding.Top);
            var Padding2 = new Point(Padding.Right, Padding.Bottom);

            return new Rect(Padding1, (-Padding2.ToVector() + Size.ToVector()).ToPoint());
        }

        protected override Size MeasureOverride(Size AvailableSize)
        {
            var InfSize = new Size(double.PositiveInfinity, double.PositiveInfinity);

            var Client = this.GetClientArea(AvailableSize);

            var ViewingRectangleQ = this.GetViewingArea(AvailableSize);
            if (!ViewingRectangleQ.HasValue)
                return AvailableSize;
            var ViewingRectangle = ViewingRectangleQ.Value;

            var Sz0 = new Point();

            Sz0 = ViewingRectangle.ToLocal01(Sz0);
            Sz0 = Client.FromLocal01(Sz0);

            foreach (UIElement C in this.Children)
            {
                var Sz = new Vector(GetW(C), GetH(C));

                if (Sz == new Vector())
                    C.Measure(InfSize);
                else
                {
                    var Szz = Sz.ToPoint();

                    Szz = ViewingRectangle.ToLocal01(Szz);
                    Szz = Client.FromLocal01(Szz);

                    var Size = (Szz - Sz0).ToSize();
                    if (double.IsNaN(Size.Width) | double.IsNaN(Size.Height))
                        Size = new Size();

                    C.Measure(Size);
                }
            }

            return new Size();
        }

        protected override Size ArrangeOverride(Size FinalSize)
        {
            var Client = this.GetClientArea(FinalSize);

            var ViewingRectangleQ = this.GetViewingArea(FinalSize);
            if (!ViewingRectangleQ.HasValue)
                return FinalSize;
            var ViewingRectangle = ViewingRectangleQ.Value;

            foreach (UIElement C in this.Children)
            {
                var Pt = new Point(GetX(C), GetY(C));
                var Sz = new Vector(GetW(C), GetH(C));

                var Pt1 = Pt;
                Pt1 = ViewingRectangle.ToLocal01(Pt1);
                Pt1 = Client.FromLocal01(Pt1);

                if (Sz == new Vector())
                    C.Arrange(new Rect(Pt1, C.DesiredSize));
                else
                {
                    var Pt2 = Pt + Sz;
                    Pt2 = ViewingRectangle.ToLocal01(Pt2);
                    Pt2 = Client.FromLocal01(Pt2);
                    C.Arrange(new Rect(Pt1, Pt2));
                }
            }

            return FinalSize;
        }

        // ToDo Support Padding and MaxViewingRectangle.

        public static readonly DependencyProperty ViewingRectangleProperty = DependencyProperty.Register("ViewingRectangle", typeof(Rect), typeof(ResizableCanvas), new PropertyMetadata(new Rect(0, 0, 1, 1), ViewingRectangle_Changed));

        private static void ViewingRectangle_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (ResizableCanvas) D;
            Self.InvalidateMeasure();
        }

        public Rect ViewingRectangle
        {
            get
            {
                return (Rect) this.GetValue(ViewingRectangleProperty);
            }
            set
            {
                this.SetValue(ViewingRectangleProperty, value);
            }
        }

        // #Region "MaxViewingRectangle Property"
        // Public Shared ReadOnly MaxViewingRectangleProperty As DependencyProperty = DependencyProperty.Register("MaxViewingRectangle", GetType(Rect?), GetType(ResizableCanvas), New PropertyMetadata(New Rect?(), AddressOf MaxViewingRectangle_Changed))

        // Private Shared Sub MaxViewingRectangle_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
        // Dim Self = DirectCast(D, ResizableCanvas)

        // Dim OldValue = DirectCast(E.OldValue, Rect?)
        // Dim NewValue = DirectCast(E.NewValue, Rect?)
        // End Sub

        // Public Property MaxViewingRectangle As Rect?
        // Get
        // Return DirectCast(Me.GetValue(MaxViewingRectangleProperty), Rect?)
        // End Get
        // Set(ByVal value As Rect?)
        // Me.SetValue(MaxViewingRectangleProperty, value)
        // End Set
        // End Property
        // #End Region

        public static readonly DependencyProperty KeepAspectRatioProperty = DependencyProperty.Register("KeepAspectRatio", typeof(bool), typeof(ResizableCanvas), new PropertyMetadata(true, KeepAspectRatio_Changed));

        private static void KeepAspectRatio_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (ResizableCanvas) D;
            Self.InvalidateMeasure();
        }

        public bool KeepAspectRatio
        {
            get
            {
                return (bool) this.GetValue(KeepAspectRatioProperty);
            }
            set
            {
                this.SetValue(KeepAspectRatioProperty, value);
            }
        }

        public static readonly DependencyProperty PaddingProperty = DependencyProperty.Register("Padding", typeof(Thickness), typeof(ResizableCanvas), new PropertyMetadata(new Thickness(), Padding_Changed));

        private static void Padding_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (ResizableCanvas) D;
            Self.InvalidateMeasure();
        }

        public Thickness Padding
        {
            get
            {
                return (Thickness) this.GetValue(PaddingProperty);
            }
            set
            {
                this.SetValue(PaddingProperty, value);
            }
        }

        public static readonly DependencyProperty YProperty = DependencyProperty.RegisterAttached("Y", typeof(double), typeof(ResizableCanvas), new PropertyMetadata(0.0, Y_Changed));

        private static void Y_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var C = VisualTreeHelper.GetParent(D) as ResizableCanvas;
            if (C != null)
                C.InvalidateMeasure();
        }

        public static double GetY(UIElement O)
        {
            return (double) O.GetValue(YProperty);
        }

        public static void SetY(UIElement O, double Value)
        {
            O.SetValue(YProperty, Value);
        }

        public static readonly DependencyProperty XProperty = DependencyProperty.RegisterAttached("X", typeof(double), typeof(ResizableCanvas), new PropertyMetadata(0.0, X_Changed));

        private static void X_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var C = VisualTreeHelper.GetParent(D) as ResizableCanvas;
            if (C != null)
                C.InvalidateMeasure();
        }

        public static double GetX(UIElement O)
        {
            return (double) O.GetValue(XProperty);
        }

        public static void SetX(UIElement O, double Value)
        {
            O.SetValue(XProperty, Value);
        }

        public static readonly DependencyProperty HProperty = DependencyProperty.RegisterAttached("H", typeof(double), typeof(ResizableCanvas), new PropertyMetadata(0.0, H_Changed));

        private static void H_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var C = VisualTreeHelper.GetParent(D) as ResizableCanvas;
            if (C != null)
                C.InvalidateMeasure();
        }

        public static double GetH(UIElement O)
        {
            return (double) O.GetValue(HProperty);
        }

        public static void SetH(UIElement O, double Value)
        {
            O.SetValue(HProperty, Value);
        }

        public static readonly DependencyProperty WProperty = DependencyProperty.RegisterAttached("W", typeof(double), typeof(ResizableCanvas), new PropertyMetadata(0.0, W_Changed));

        private static void W_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var C = VisualTreeHelper.GetParent(D) as ResizableCanvas;
            if (C != null)
                C.InvalidateMeasure();
        }

        public static double GetW(UIElement O)
        {
            return (double) O.GetValue(WProperty);
        }

        public static void SetW(UIElement O, double Value)
        {
            O.SetValue(WProperty, Value);
        }
    }
}
